trie/node.go

```diff
@@ -30,42 +30,63 @@ var indices = []string{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "a", "b
 type node interface {
 	fstring(string) string
 	cache() (hashNode, bool)
+	canUnload(cachegen, cachelimit uint16) bool
 }
 
 type (
 	fullNode struct {
 		Children [17]node // Actual trie node data to encode/decode (needs custom encoder)
-		hash     hashNode // Cached hash of the node to prevent rehashing (may be nil)
-		dirty    bool     // Cached flag whether the node's new or already stored
+		nodeFlag
 	}
 	shortNode struct {
-		Key   []byte
-		Val   node
-		hash  hashNode // Cached hash of the node to prevent rehashing (may be nil)
-		dirty bool     // Cached flag whether the node's new or already stored
+		Key []byte
+		Val node
+		nodeFlag
 	}
 	hashNode  []byte
 	valueNode []byte
 )
 
 // EncodeRLP encodes a full node into the consensus RLP format.
-func (n fullNode) EncodeRLP(w io.Writer) error {
+func (n *fullNode) EncodeRLP(w io.Writer) error {
 	return rlp.Encode(w, n.Children)
 }
 
-// Cache accessors to retrieve precalculated values (avoid lengthy type switches).
-func (n fullNode) cache() (hashNode, bool)  { return n.hash, n.dirty }
-func (n shortNode) cache() (hashNode, bool) { return n.hash, n.dirty }
+func (n *fullNode) copy() *fullNode   { copy := *n; return &copy }
+func (n *shortNode) copy() *shortNode { copy := *n; return &copy }
+
+// nodeFlag contains caching-related metadata about a node.
+type nodeFlag struct {
+	hash  hashNode // cached hash of the node (may be nil)
+	gen   uint16   // cache generation counter
+	dirty bool     // whether the node has changes that must be written to the database
+}
+
+// canUnload tells whether a node can be unloaded.
+func (n nodeFlag) canUnload(cachegen, cachelimit uint16) bool {
+	var dist uint16
+	if n.gen > cachegen {
```

Comment:
This if statement should not be here. Suppose `n.gen = 65534`; when `cachegen = 65535`, you calculate (correctly) `cachegen - n.gen = 65535 - 65534 = 1`. When `cachegen = 0` you calculate (incorrectly) `n.gen - cachegen = 65534 - 0 = 65534`.

If instead you always compute `cachegen - n.gen`, the result in the last case is `0 - 65534 = -65534 = 2 mod 65536`.

Are you able to add a test for integer overflow?